test/replication/gh-5430-cluster-mvcc.test.lua

```diff
@@ -37,16 +37,18 @@ box.error.injection.set('ERRINJ_WAL_DELAY', true)
 test_run:switch('replica1')
 _ = require('fiber').create(function() box.cfg{replication_anon = false} end)
 str = string.format('registering replica %s', box.info.uuid):gsub('-', '%%-')
-_ = test_run:wait_log('master', str)
+test_run:wait_log('master', str) ~= nil
 
 test_run:switch('replica2')
 _ = require('fiber').create(function() box.cfg{replication_anon = false} end)
-str = string.format('registering replica %s', box.info.uuid):gsub('-', '%%-')
-_ = test_run:wait_log('master', str)
 
 test_run:switch('master')
 box.error.injection.set('ERRINJ_WAL_DELAY', false)
 
+test_run:switch('replica2')
+str = string.format('registering replica %s', box.info.uuid):gsub('-', '%%-')
+test_run:wait_log('master', str) ~= nil
+
 test_run:switch('replica1')
```

Comment:
The test was supposed to get 2 concurrent `_cluster` txns running and then unblock WAL. But now the test might be starting the registration of `replica1` on the master, then finishing the WAL delay, and then starting the registration of `replica2`. Because you start the latter in a fiber without waiting for anything. It could as well not reach the master until you make the second `wait_log` after the first replica is already registered.

Why was `wait_log` timing out?

Reply:
Sorry, I misunderstood the purpose of the test.

It seems like after 70a68836 we are unable to register an anonymous replica while its relay is not finished:
https://github.com/tarantool/tarantool/blob/9e616ab6f2f94ba2bcad3044187c474a1c05a8c5/src/box/box.cc#L4422-L4427

At the same time, the anonymous replica's relay cannot finish, because it is blocked by `cbus_unpair` in `wal_clear_watcher`:
https://github.com/tarantool/tarantool/blob/9e616ab6f2f94ba2bcad3044187c474a1c05a8c5/src/box/relay.cc#L1068-L1074

https://github.com/tarantool/tarantool/blob/9e616ab6f2f94ba2bcad3044187c474a1c05a8c5/src/lib/core/cbus.h#L440-L442

Reply:
Hm. The case is definitely still valid, but the test indeed needs a rework. Since this is MVCC, I would carefully assume, that instead of box.error.injection.set('ERRINJ_WAL_DELAY', true) we could set an on-replace trigger for _cluster, which would yield until allowed to continue. That way we could get 2 txns on _cluster in parallel.

If this won't work somehow, I suggest to move ERRINJ_REPLICA_JOIN_DELAY before box_register_replica(). This errinj is not used right now for anything, so we can reuse it here instead of ERRINJ_WAL_DELAY. Then WAL isn't blocked, cbus_unpair with WAL endpoints will work, and yet the replicas won't register in turns.

Once we see that both of them are registering replica in the logs, we can set ERRINJ_WAL_DELAY and unset ERRINJ_REPLICA_JOIN_DELAY.

Reply:
Thanks for the help!

We cannot do it the first way because it is forbidden to yield during system space changes.

But the second way seems to be okay, please check the test changes.